shell/platform/windows/window_top_level_message_handler.cc

```diff
@@ -0,0 +1,74 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "window_top_level_message_handler.h"
+
+#include <TlHelp32.h>
+#include <Windows.h>
+#include <WinUser.h>
+#include <tchar.h>
+
+#include "flutter/shell/platform/windows/flutter_windows_engine.h"
+
+namespace flutter {
+
+static constexpr char kExitTypeCancelable[] = "cancelable";
+
+WindowTopLevelMessageHandler::WindowTopLevelMessageHandler(FlutterWindowsEngine& engine) : engine_(engine) {}
+
+WindowTopLevelMessageHandler::~WindowTopLevelMessageHandler() {}
+
+void WindowTopLevelMessageHandler::Quit(int64_t exit_code) {
+  PostQuitMessage(exit_code);
+}
+
+bool WindowTopLevelMessageHandler::WindowProc(HWND hwnd, UINT msg, WPARAM wpar, LPARAM lpar, LRESULT* result) {
+  switch (msg) {
+    case WM_CLOSE:
+      if (IsLastWindowOfProcess()) {
+        engine_.RequestApplicationQuit(kExitTypeCancelable, wpar);
+      }
+      return true;
+  }
+  return false;
+}
+
+static BOOL CALLBACK WindowEnumCallback(HWND hwnd, LPARAM user_data) {
```

Comment:
Could you call `GetParent` on the window here and only count the ones that return NULL?